gcovr/configuration.py

```diff
@@ -0,0 +1,475 @@
+    representing these options as a simple class object allows them to be
+    portabilty re-used in multiple configuration schemes. This is implemented
+    in a way similar to how options are defined in argparse. The converter
+    keyword argument is expected to return a valid conversion of a string
+    value or throw an error.
+
+    This class is initializer in a very similar fashion as adding arguments
+    to an argparse Parser instance with a few minor changes.
+        1. The first positional argument is the "name" (i.e. "dest" in
+           argparse speak) and is required.
+        2. A "group" keyword argument has been added which corresponds to a
+           key in the GCOVR_CONFIG_OPTION_GROUPS dictionary.
+    """
+
+    def __init__(self, name, *flags, **kwargs):
+        self.name = name
+        self.flags = flags
+        self.group = kwargs.pop("group", None)
+        self.default = None
+        #
+        # set remaining kwargs as attributes
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+        if hasattr(self, 'help'):
+            self.help = self.help.format(**self.__dict__)
+        #
+        # the store_true and store_false actions have hardcoded boolean
+        # constants in their definitions so they need switched to the generic
+        # store_const in order for the logic here to work correctly.
+        if getattr(self, 'action', None) == 'store_true':
+            self.action = 'store_const'
+            self.const = True
+            self.default = False
+        if getattr(self, 'action', None) == 'store_false':
+            self.action = 'store_const'
+            self.const = False
+            self.default = True
+
+    def __repr__(self):
+        r"""String representation of instance."""
+        fmt = "GcovrConfigOption({name!r}, {flags}group={group!r}, {kwargs})"
+        kwargs = dict(self.__dict__)
+        for key in ["name", "flags", "group"]:
+            del kwargs[key]
+        kwargs = ['{k}={v!r}'.format(k=k, v=v) for k, v in kwargs.items()]
+        #
+        return fmt.format(name=self.name,
+                          flags=', '.join(self.flags) + ', ',
+                          group=self.group,
+                          kwargs=', '.join(kwargs))
+
+    def add_option_to_parser(self, parser):
+        r"""Adds the argument to an argparse parser or option group."""
+        # extract keyword args from instance
+        keys = ["action", "nargs", "const", "type", "choices",
+                "required", "help", "metavar"]
+        kwargs = {"default": SUPPRESS}
+        for key in keys:
+            if hasattr(self, key):
+                kwargs[key] = getattr(self, key)
+
+        # we only want to set dest for non-positionals
+        if self.flags:
+            kwargs["dest"] = self.name
+            parser.add_argument(*self.flags, **kwargs)
+        else:
+            parser.add_argument(self.name, **kwargs)
+
+
+GCOVR_CONFIG_OPTION_GROUPS = {
```

Comment:
dict order is undefined prior to Python 3.6. We can use `collections.OrderedDict` here so that the groups are always in the correct order